include/drivers/reset.h

```diff
@@ -0,0 +1,192 @@
+ * @brief Public Reset Controller driver APIs
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_RESET_H_
+#define ZEPHYR_INCLUDE_DRIVERS_RESET_H_
+
+/**
+ * @brief Reset Controller Interface
+ * @defgroup reset_controller_interface Reset Controller Interface
+ * @ingroup io_interfaces
+ * @{
+ */
+
+#include <zephyr/types.h>
+#include <device.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/** Reset controller device configuration. */
+struct reset_dev_config {
+	/** Reset controller device. */
+	const struct device *dev;
+	/** Reset line. */
+	uint32_t id;
+};
+
+/**
+ * @brief Helper macro to initialize reset controller device config.
+ *
+ * @param inst Instance number.
+ * @param id Reset line.
+ */
+#define RESET_DT_INST_DEV_CONFIG_INIT(inst, id)			\
+	{							\
+		.dev = DEVICE_DT_GET(DT_RESET_INST_CTLR(inst)),	\
+		.id = DT_RESET_INST_CELL(inst, id),		\
+	}
+
+/** @cond INTERNAL_HIDDEN */
+
+/**
+ * API template to get the reset status of the device.
+ *
+ * @see reset_status
+ */
+typedef int (*reset_api_status)(const struct device *dev, uint32_t id, uint8_t *status);
+
+/**
+ * API template to put the device in reset state.
+ *
+ * @see reset_assert
+ */
+typedef int (*reset_api_assert)(const struct device *dev, uint32_t id);
+
+/**
+ * API template to take out the device from reset state.
+ *
+ * @see reset_deassert
+ */
+typedef int (*reset_api_deassert)(const struct device *dev, uint32_t id);
+
+/**
+ * API template to reset the device.
+ *
+ * @see reset_toggle
+ */
+typedef int (*reset_api_toggle)(const struct device *dev, uint32_t id);
+
+/**
+ * @brief Reset Controller driver API
+ */
+__subsystem struct reset_driver_api {
+	reset_api_status status;
+	reset_api_assert assert;
+	reset_api_deassert deassert;
+	reset_api_toggle toggle;
+};
+
+/** @endcond */
+
+/**
+ * @brief Get the reset status
+ *
+ * This function returns the reset status of the device.
+ *
+ * @param reset Reset controller device configuration.
+ * @param status Where to write the reset status.
+ *
+ * @return 0 on success, negative code on error.
+ */
+__syscall int reset_status(const struct reset_dev_config *reset, uint8_t *status);
+
+static inline int z_impl_reset_status(const struct reset_dev_config *reset, uint8_t *status)
+{
+	const struct reset_driver_api *api = (const struct reset_driver_api *)reset->dev->api;
+
+	if (!api) {
+		return -ENOTSUP;
+	}
+
+	return api->status(reset->dev, reset->id, status);
+}
+
+/**
+ * @brief Put the device in reset state
+ *
+ * This function sets/cleares the reset bits of the device,
+ * depending on the logic level (active-high/active-low).
+ *
+ * @param reset Reset controller device configuration.
+ *
+ * @return 0 on success, negative code on error.
+ */
+__syscall int reset_assert(const struct reset_dev_config *reset);
+
+static inline int z_impl_reset_assert(const struct reset_dev_config *reset)
+{
+	const struct reset_driver_api *api = (const struct reset_driver_api *)reset->dev->api;
+
+	if (!api) {
+		return -ENOTSUP;
```

Comment:
`-ENOSYS`

Reply:
Can I add also some checks like if (!api->assert) { return -ENOSYS; }?

Reply:
Sorry, I realized now that this is de-referencing twice. I think we should do then:

1. An API that just takes a `const struct device *` + `uint32_t id`
2. A "_dt" API that takes a sort of `_dt_spec` config (see for example GPIO or I2C APIs, they use this pattern)

There's no need to check for NULL API, I think. I misunderstood the code and thought that the API was allowed to not implement the call.

Reply:
please check for the `_dt` approach, as other APIs like GPIO/I2C/SPI do.

https://github.com/zephyrproject-rtos/zephyr/blob/main/include/drivers/i2c.h#L67
https://github.com/zephyrproject-rtos/zephyr/blob/main/include/drivers/i2c.h#L82
https://github.com/zephyrproject-rtos/zephyr/blob/main/include/drivers/i2c.h#L609